### DISCUSSION TITLE
Get Application Audio Sessions

### DISCUSSION THREAD

**Discussion by seanboehnke** [naudio/NAudio#1144] [Q&A]
I'm working on an application that needs to be able to get individual audio data from applications, filter specific ones out based on which application the audio comes from, and then combine the resulting audio sources into one stream for network streaming. Would NAudio be able to cover this application's needs?

**Comment by JavierparraDev** [comment]
Short answer: NAudio covers the **session enumeration** and **mixing** parts, but capturing the raw PCM of a *single* application is not something NAudio wraps today — you need Windows' **Process Loopback** API (Windows 10 2004+) or a virtual cable per app.

**1. Enumerating per-app sessions (yes, NAudio)**

`MMDeviceEnumerator` + `AudioSessionManager` gives you per-application identity, volume, mute and peak level:

```csharp
using var enumerator = new MMDeviceEnumerator();
var device = enumerator.GetDefaultAudioEndpoint(DataFlow.Render, Role.Multimedia);
var sessions = device.AudioSessionManager;
sessions.RefreshSessions();

for (int i = 0; i < sessions.Sessions.Count; i++)
{
    var session = sessions.Sessions[i];
    var pid   = session.GetProcessID;                 // identify the app
    var state = session.State;                        // Active / Inactive / Expired
    var peak  = session.AudioMeterInformation.MasterPeakValue;
}
```

This is enough to *know* which app is playing, but it does not give you the audio samples.

**2. Capturing a single app's audio (Process Loopback)**

Since Windows 10 2004, `ActivateAudioInterfaceAsync` supports `AUDIOCLIENT_ACTIVATION_TYPE_PROCESS_LOOPBACK`. You pass an `AUDIOCLIENT_ACTIVATION_PARAMS` whose `ProcessLoopbackParams` contain the target PID and `PROCESS_LOOPBACK_MODE_INCLUDE_TARGET_PROCESS_TREE`, then read frames from the resulting `IAudioClient` in loopback mode. NAudio does not ship a wrapper, so you implement the COM interop yourself. Caveats: the process must be actively rendering, exclusive-mode streams are excluded, and protected content cannot be captured.

**3. Filtering and mixing**

Once each capture yields a byte stream, wrap each one in a `BufferedWaveProvider` and combine them with `MixingSampleProvider` (or sum the samples yourself). To filter an app out, simply don't add its provider to the mix.

**4. Alternative without per-process capture**

Route each app to its own virtual cable (VB-CABLE, Voicemeeter, …) and capture each cable as a normal loopback device. This is how most shipping tools do it: no exotic APIs, but the user must install a virtual device.

If it helps as a reference, [AudioFlow](https://github.com/JavierparraDev/audioflow) is an MIT-licensed Windows per-app audio router built on NAudio. The session enumeration lives in `AudioFlow.Core/WindowsAudio` and the Process Loopback probe in `AudioFlow.ProcessLoopback`, so you can see both approaches in context.
